design/restic-backup-and-restore-progress.md

```diff
@@ -0,0 +1,104 @@
+```
+
+The [command factory for backup](https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/pkg/restic/command_factory.go#L37) will be updated to include the `--json` flag.
+The code to run the `restic backup` command (https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/pkg/controller/pod_volume_backup_controller.go#L241) will be changed to include a Goroutine that reads from the command's stdout stream.
+The implementation of this will largely follow [@jmontleon's PoC](https://github.com/fusor/velero/pull/4/files) of this.
+The Goroutine will periodically read the stream (every 10 seconds) and get the last printed status line.
+The line will be decoded as JSON and the `percent_done` property will be read and formatted as a percentage value.
+
+The PodVolumeBackupStatus type will be extended to include a `Progress` field.
+The PodVolumeBackup will be patched to set the `status.Progress` field to the percentage value of the progress.
+
+Once the backup has completed successfully, the PodVolumeBackup will be patched to set `status.Progress = 100%`.
+
+### restic restore progress
+
+The `restic ls <snapshot_id> --json` command provides information about the size of backed up files:
+
+```
+{"time":"2019-08-14T15:16:29.105557-07:00","tree":"1de5381473aecb5548959f6e5444bb7966fe781500b5f8ec1629088674bea6a6","paths":["/Users/aadnan/Playground/restic/files"],"hostname":"aadnan-a01.vmware.com","username":"aadnan","uid":501,"gid":20,"id":"5628926b4f0e105eb33d83c619f235a8cf8f695d19955fab44bf506ece195f3a","short_id":"5628926b","struct_type":"snapshot"}
+{"name":"files","type":"dir","path":"/files","uid":501,"gid":20,"mode":2147484096,"mtime":"2019-08-14T15:16:20.53091354-07:00","atime":"2019-08-14T15:16:20.53091354-07:00","ctime":"2019-08-14T15:16:20.53091354-07:00","struct_type":"node"}
+{"name":"file1","type":"file","path":"/files/file1","uid":501,"gid":20,"size":1435500544,"mode":384,"mtime":"2019-08-14T15:16:20.491206506-07:00","atime":"2019-08-14T15:16:20.491206506-07:00","ctime":"2019-08-14T15:16:20.491234451-07:00","struct_type":"node"}
+...
+```
+
+Before beginning the restore operation, we can use the output of `restic ls` to get a list of the files that will be restored and calculate the total size of the backup.
+
+The code to run the `restic restore` command will be changed to include a Goroutine that periodically (every 10 seconds) goes through the list of files in the backup and gets the current size of each file in the volume.
+If the file doesn't exist in the volume yet (restic hasn't started restoring it), the size is 0.
+The sum of the current size of each file is then compared with the total size of the backup to calculate a completion percentage (`cursize / totalsize * 100`).
+
+The PodVolumeRestoreStatus type will be extended to include a `Progress` field.
```

Comment:
similarly, I'd go with `TotalBytes` and `BytesComplete` and calculate percent complete in UI code.